File: 자료구조/Tree.java
package 자료구조;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;

public class Tree{
    public static void main(String[] args) {
        Tree_ t = new Tree_("A");
        t.add_node("A", "B");
        t.add_node("A", "C");
        t.add_node("B", "D");
        t.add_node("C", "F");

        System.out.print(t.root.data + " -> ");
        for(Node i : t.root.children){
            System.out.print(i.data+ " ");
        }
        System.out.println();

        Node b = t.root.children.get(0);
        Node c = t.root.children.get(1);
        
        System.out.print(b.data+ " -> ");
        for(Node i : b.children){
            System.out.print(i.data + " ");
        }
        System.out.println();

        System.out.print(c.data+ " -> ");

        for(Node i : c.children){
            System.out.print(i.data + " ");
        }
    }
}
class Tree_ {
    Node root;
    
    Tree_(String data) {
        this.root = new Node(data);
    }

    public void add_node(String targetNode, String data){
        Deque<Node> q = new LinkedList<>();
        q.addLast(this.root);
        while(!q.isEmpty()){
            Node currentNode = q.pollFirst();
            if (currentNode.data == targetNode) {
                currentNode.children.add(new Node(data));
                break;
            }
            
            for(Node i : currentNode.children){
                q.addLast(i);
            }

        }
    }
}

class Node{
    public String data;
    public ArrayList<Node> children;

    Node(String data){
        this.data = data;
        this.children = new ArrayList<>();
    }
}

File: 자료구조/Tree.py
from collections import deque

class Node:
    def __init__(self, data) -> None:
        self.data = data
        self.childen = []
    
class Tree:
    def __init__(self, data) -> None:
        self.root = Node(data)
    
    def add_node(self, target_data, new_data, ) -> None:
        q = deque()
        q.append(self.root)
        while q:
            current_node = q.popleft()
            if current_node.data == target_data:
                current_node.childen.append(Node(new_data))        
                break
            
            for i in current_node.childen:
                q.append(i)
    
tree = Tree("A")
tree.add_node("A", "B")
tree.add_node("A", "C")
tree.add_node("B", "D")
tree.add_node("C", "F")

print("A의 자식 :", end=" ")
for i in tree.root.childen:
    print(i.data, end=" ")
print("\nB의 자식 :", end=" ")
for i in tree.root.childen[0].childen:
    print(i.data, end=" ")    
print("\nC의 자식 :", end=" ")
for i in tree.root.childen[1].childen:
    print(i.data, end=" ")
